resources/net.go

```diff
@@ -0,0 +1,849 @@
+				routeOK = true
+				break
+			}
+		}
+		// if any addr is missing a kernel route return early
+		if !routeOK {
+			break
+		}
+	}
+	return routeOK, nil
+}
+
+// kernelApply adds or replaces each address' kernel route as necessary.
+func (i *iface) kernelApply(addrs []string) error {
+	// for each addr, add or replace the corresponding kernel route
+	for _, addr := range addrs {
+		ip, ipNet, err := net.ParseCIDR(addr)
+		if err != nil {
+			return errwrap.Wrapf(err, "error parsing addr: %s", addr)
+		}
+		// kernel route needed for the network to be reachable from a given ip
+		if err := netlink.RouteReplace(&netlink.Route{
+			LinkIndex: i.Index,
+			Dst:       ipNet,
+			Src:       ip,
+			Protocol:  rtProtoKernel,
+			Scope:     netlink.SCOPE_LINK,
+		}); err != nil {
+			return errwrap.Wrapf(err, "error replacing first hop route")
+		}
+	}
+	return nil
+}
+
+// addrApplyDelete, checks the interface's addresses and deletes any that are not
+// in the list/definition.
+func (i *iface) addrApplyDelete(objAddrs []string) error {
+	ifaceAddrs, err := i.getAddrs()
+	if err != nil {
+		return errwrap.Wrapf(err, "error getting addrs from interface: %s", i.Name)
+	}
+	for _, ifaceAddr := range ifaceAddrs {
+		addrOK := false
+		for _, objAddr := range objAddrs {
+			if ifaceAddr == objAddr {
+				addrOK = true
+			}
+		}
+		if addrOK {
+			continue
+		}
+		addr, err := netlink.ParseAddr(ifaceAddr)
+		if err != nil {
+			return errwrap.Wrapf(err, "error parsing netlink address: %s", ifaceAddr)
+		}
+		if err := netlink.AddrDel(i.link, addr); err != nil {
+			return errwrap.Wrapf(err, "error deleting addr: %s from %s", ifaceAddr, i.Name)
+		}
+	}
+	return nil
+}
+
+// addrApplyAdd checks if the interface has each address in the supplied list,
+// and if it doesn't, it adds them.
+func (i *iface) addrApplyAdd(objAddrs []string) error {
+	ifaceAddrs, err := i.getAddrs()
+	if err != nil {
+		return errwrap.Wrapf(err, "error getting addrs from interface: %s", i.Name)
+	}
+	for _, objAddr := range objAddrs {
+		addrOK := false
+		for _, ifaceAddr := range ifaceAddrs {
+			if ifaceAddr == objAddr {
+				addrOK = true
+			}
+		}
+		if addrOK {
+			continue
+		}
+		addr, err := netlink.ParseAddr(objAddr)
+		if err != nil {
+			return errwrap.Wrapf(err, "error parsing cidr address: %s", objAddr)
+		}
+		if err := netlink.AddrAdd(i.link, addr); err != nil {
+			return errwrap.Wrapf(err, "error adding addr: %s to %s", objAddr, i.Name)
+		}
+	}
+	return nil
+}
+
+// fdSetAdd adds the given file descriptor to the FdSet. FdSet is a bitmask
+// representing the file descriptors that unix.Select will use as input.
+func fdSetAdd(fdSet *unix.FdSet, fd int) {
+	fdSet.Bits[fd/64] |= 1 << uint(fd) // fd = 3 becomes 100 if we add 5, we get 10100
+}
+
+// strSliceCompare takes two lists of strings and returns whether or not they
+// are identical.
+func strSliceCompare(x, y []string) bool {
+	if len(x) != len(y) {
+		return false
+	}
+	sort.Strings(x)
```

Comment:
I am stupid and forget, but doesn't this modify the input args? Check it doesn't have a side effect please :)

Also your docs are not exactly correct for this function if you sort the two before compare.

Lastly, please add this function to util/util.go package with a test (at least testing the input modification thing) or two please :) Signature should be: `x, y []string) error` instead.

Reply:
You are correct. We discussed this, and I forgot to change it. Coming from Haskell/Rust this seems like something the language shouldn't do without a pointer reference. But, what do I know?
In what case should this return an error (instead of false?)

Reply:
if we return bool, we have two possible values. if we return error we have two possible values (nil vs. err) but in the second case we can provide a *reason* which is more cool in case anyone cares.